### PR TITLE
Support legacy create asset servlet

### LINKED ISSUE #131: Support legacy create asset servlet (i.e. non-cloud service instances)
### Expected Behaviour

AEM Desktop (a consumer of this library) supports AEM instances all the way back to 6.4. This version of AEM doesn't include direct binary access, so asset uploads are done using the create asset servlet. This servlet supports uploading in parts, but acts as a more "standard" POST-like endpoint, where requests sent to it include a binary's data in the body.

The desktop app supports this by maintaining two "modes":

* Mode 1 is active against AEM instances that have direct binary access enabled, and uses this library to transfer local binaries to AEM.
* Mode 2 is active when direct binary access is _not_ enabled, and uses a less efficient, home-grown solution for uploading to the create asset servlet.

Upon revisiting mode 2, I started wondering how intrusive it would be to support the create asset servlet with this module. After some analysis and prototyping, I discovered that it was minimally intrusive, and actually fairly simple to extend the transfer process to support the create asset servlet.

By including this feature here, we can greatly simplify a lot of the desktop app's code, because we can entirely remove mode 2 and centralize on using this library for uploading assets to AEM, no matter the version.

### Actual Behaviour

This library only works with AEM instances that are configured with direct binary access.

### Reproduce Scenario (including but not limited to)

Attempting to use this library to upload assets to an AEM instance that doesn't have direct binary access enabled.

#### Steps to Reproduce

Attempt to use `AEMUpload` to upload assets to an AEM instance without direct binary access enabled. Note that the upload fails with an error.

#### Platform and Version

Reproducible on any platform or version.

#### Sample Code that illustrates the problem

N/A

#### Logs taken while reproducing problem

N/A

### PR DESCRIPTION
**Pull request by mfrisbey**
## Description

Ok, hear me out.

See linked issue for more context, but I discovered that it wasn't super difficult for this library to work with the old create asset servlet. By supporting this, we can update `aem-upload` so that it can upload binaries to _any_ version of AEM. This will trickle down to AEM Desktop and allow us to remove quite a bit of legacy code.

In summary, here is what this PR does:

* Adds a new class, `CreateAssetServletUpload`, which can be used similar to `AEMUpload` to upload assets; the new upload process uses the create asset servlet instead of the direct binary upload algorithm.
* Adds a couple new methods to `TransferPart` and `TransferProcess` that allow new child classes to override behavior in a couple of key areas. This allows the transfer step that normally goes to Azure URLs to work with the create asset servlet, too.
* Adds unit tests for the new classes.
* Adds new e2e tests for verifying uploading with the create asset servlet. This does change the e2e tests so that they require two AEM environments: one with direct binary access enabled, and one without.

One thing to note is that concurrency doesn't work well with the create asset servlet, so even though the transfer process will still break a file into parts, the parts will always be uploaded sequentially to the create asset servlet.

Fixes #131 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: lib/asset/createassetservletpart.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
```
Comment (tmathern) [inline lib/asset/createassetservletpart.js:2]:
```suggestion
 * Copyright 2022 Adobe. All rights reserved.
```

File: lib/asset/createassetservletpart.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+const UTF8 = "utf-8";
+const CHUNKED_CONTENT_TYPE = "x-chunked-content-type";
+const CHUNKED_TOTAL_SIZE = "x-chunked-total-size";
+
+/**
+ * Reflects a unit of content to be transferred, refers to part of an asset. This
+ * part provides functionality specific to uploading to AEM's create asset servlet.
+ */
+class CreateAssetServletPart extends TransferPartImport.TransferPart {
+    /**
```
Comment (tmathern) [inline lib/asset/createassetservletpart.js:33]: Misses description of what the method does

File: lib/asset/createassetservletpart.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        }
+        form.append(FILE, partData, {
+            filename: this.transferAsset.target.filename,
+            [HTTP.HEADER.CONTENT_TYPE]: this.contentType
+        });
+
+        return form;
+    }
+
+    /**
```
Comment (tmathern) [inline lib/asset/createassetservletpart.js:59]: Missing description of what the method does

File: lib/asset/createassetservletpart.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+     */
+    get totalSize() {
+        return this.transferAsset.metadata.contentLength;
+    }
+
+    /**
+     * Returns the content type of the asset being transferred.
+     * @returns {string} Content type of the asset.
+     */
+    get contentType() {
```
Comment (tmathern) [inline lib/asset/createassetservletpart.js:104]: I think this method could even move in a parent class (I'd put it at the top level)

File: lib/asset/transferpart.js [hunk 5]
```diff
@@ -118,6 +132,31 @@ class TransferPart {
...
+    /**
+     * Creates the headers that will be used for the part when transferred via HTTP.
+     * @param {PartHttpInfo} partHttpInfo HTTP information about the part.
+     * @returns {Headers} Simple object containing header information.
+     */
+    createPartHttpHeaders(partHttpInfo) {
+        const { httpBody } = partHttpInfo;
+        // the body could be a stream or a blob, use correct property
+        return {
+            [HTTP.HEADER.CONTENT_LENGTH]: httpBody.length || httpBody.size,
```
Comment (tmathern) [inline lib/asset/transferpart.js:156]: Will HTTP.HEADER... always be fully set when this call happens?

File: README.md [hunk 6]
```diff
@@ -126,6 +126,19 @@ async main() {
 
 Assuming `test.png` is 1,800,000 bytes this will upload the first 1,000,000 bytes to `http://my.server.com/test.png.1` and the next 800,000 bytes to `http://my.server.com/test.png.2`.
 
+Upload files using AEM's create asset servlet:
```
[the remark this answers is not in the document]
Comment (tmathern) [reply README.md:129]: Maybe even add a support/recommended methods data matrix with the AEM version as input?

File: test/asset/transferpart.test.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2021 Adobe. All rights reserved.
```
Comment (tmathern) [inline test/asset/transferpart.test.js:2]:
```suggestion
 * Copyright 2022 Adobe. All rights reserved.
```

**Review by tmathern — COMMENTED** [review]
This adds a lot of additional classes and layers, which increase complexity and the maintainability burden. Documentation must also be updated to state preferred methods, because otherwise it's guesswork for developers starting to use it and who may not know all AEM systems details and capabilities, resulting in a somewhat bad developer experience. 

Is there data about how many AEM instances this would serve/support? 
Because eventually this may simplify AEM Desktop, but moves the complexity (currently, without refactoring) into this library which is also used by Asset Link and Asset Compute.

In any case, I strongly suggest to review and refactor this code to reduce the amount of churning through files and layers *in another PR*.

**Comment by mfrisbey** [comment]
@tmathern Thanks for the review!

> This adds a lot of additional classes and layers, which increase complexity and the maintainability burden.

That's subjective, IMO. What specifically about the solution has you concerned? Is there something overly complicated about it? I actually thought it was pretty straightfoward - by adding a couple new methods and following the existing class-based design, I was able to plug a different endpoint into the overall algorithm without sweeping refactoring of existing functionality.

You also mentioned reducing files and churn. How would you suggest doing that? Do we just add another `if` statement in a file somewhere so that we have fewer files? IMO that would make the code even _more_ complex. How many layers would be more acceptable? What's so bad about adding 4 more classes, especially when they either follow the same pattern as other existing classes, or are trivial? They've also got unit tests and extended end-to-end tests. What are your specific concerns about the complexity and maintainability here? I want to make sure I understand so that I can address them correctly.

**Comment by tmathern** [comment]
> @tmathern Thanks for the review!
> 
> > This adds a lot of additional classes and layers, which increase complexity and the maintainability burden.
> 
> That's subjective, IMO. What specifically about the solution has you concerned? Is there something overly complicated about it? I actually thought it was pretty straightfoward - by adding a couple new methods and following the existing class-based design, I was able to plug a different endpoint into the overall algorithm without sweeping refactoring of existing functionality.
> 
> You also mentioned reducing files and churn. How would you suggest doing that? Do we just add another `if` statement in a file somewhere so that we have fewer files? IMO that would make the code even _more_ complex. How many layers would be more acceptable? What's so bad about adding 4 more classes, especially when they either follow the same pattern as other existing classes, or are trivial? They've also got unit tests and extended end-to-end tests. What are your specific concerns about the complexity and maintainability here? I want to make sure I understand so that I can address them correctly.

What I am suggesting is to refactor all of this. Maybe even redesign it. Put yourself in the shoes of a developer who never worked with this library. It is a nightmare to get started with because you can't guess easily where/how. And you go through a lot of files following the code (I'm not even talking about file naming, which is neither camel nor snake case).  And to add to the confusion, I see a lot of similar code (not repeated, but similar), which makes me think there is value in refactoring here.
Using a class based design is good, but at that point instead of continuing of adding methods (and classes and layers) here and there, we should consider redesigning it to move everything that can be shared (made variable) up in the hierarchy (and make it more abstract in the upper levels) and reduce the amount of code in the children classes. I think it would make it easier, but maybe that's just me.

**Comment by mfrisbey** [comment]
> I'm not even talking about file naming, which is neither camel nor snake case

I followed the name style of all the other files. I agree that it's not great, but that's how all the other files are named. Should I deviate from the style of the project?

> I see a lot of similar code (not repeated, but similar)

Do you have examples? I haven't looked at your individual comments yet, so if you point it out there then just tell me to read through those :joy:

> It is a nightmare to get started with because you can't guess easily where/how

Is that an issue with the code itself, or would some improved documentation solve that?

> What I am suggesting is to refactor all of this. Maybe even redesign it. 

When you say "this", are you talking about the code that I contributed, or the class design of the module in general? If you're talking about this code specifically, I could certainly take another pass at it. But I still maintain that this code isn't overly complex, especially considering the already complex nature of the module itself.

I could also certainly take a more "destructive" approach and refactor more of the existing code so that I can fit these changes in with fewer classes/files. But then we're running the risk of breaking things by touching more code. IMO this change is low risk because it's minimally invasive with the surrounding code.

Again, I'm happy to take another pass at it, I'm just going to need more suggestions and feedback about what makes it such a nightmare to follow. But what I _would_ like to avoid is an invasive change to support this feature. I don't want to drastically alter the library to add this code, because it's not worth the risk. The major reason I even moved forward with the proposal is because I felt it fit in nicely with the existing code, and has a low risk of causing regressions.

**Comment by tmathern** [comment]
> > I'm not even talking about file naming, which is neither camel nor snake case
> 
> I followed the name style of all the other files. I agree that it's not great, but that's how all the other files are named. Should I deviate from the style of the project?
> 
> > I see a lot of similar code (not repeated, but similar)
> 
> Do you have examples? I haven't looked at your individual comments yet, so if you point it out there then just tell me to read through those 😂
> 
> > It is a nightmare to get started with because you can't guess easily where/how
> 
> Is that an issue with the code itself, or would some improved documentation solve that?
> 
> > What I am suggesting is to refactor all of this. Maybe even redesign it.
> 
> When you say "this", are you talking about the code that I contributed, or the class design of the module in general? If you're talking about this code specifically, I could certainly take another pass at it. But I still maintain that this code isn't overly complex, especially considering the already complex nature of the module itself.
> 
> I could also certainly take a more "destructive" approach and refactor more of the existing code so that I can fit these changes in with fewer classes/files. But then we're running the risk of breaking things by touching more code. IMO this change is low risk because it's minimally invasive with the surrounding code.
> 
> Again, I'm happy to take another pass at it, I'm just going to need more suggestions and feedback about what makes it such a nightmare to follow. But what I _would_ like to avoid is an invasive change to support this feature. I don't want to drastically alter the library to add this code, because it's not worth the risk. The major reason I even moved forward with the proposal is because I felt it fit in nicely with the existing code, and has a low risk of causing regressions.

First step is to improve the doc. I suggested some ideas to improve that. 

I would defer the refactoring changes for now, but keep that in mind, and it's for the library/module in general as it grows, not this addition only. Taking a step back to improve.  

An example of repetitive/refactorable code is the similarity of steps https://github.com/adobe/node-httptransfer/blob/70d65e4b8cef52ee99d4ddbebe4fdc15a82d0625/lib/aem/aemupload.js and https://github.com/adobe/node-httptransfer/blob/70d65e4b8cef52ee99d4ddbebe4fdc15a82d0625/lib/aem/createassetservletupload.js follow to prepare the upload. Eg. I think we could express those steps somewhere, and make them modular. But that needs analysis and good testing, hence suggesting to do it in another PR. 

But I would love to hear also other opinions on this (especially @bobvanmanen's). 

To finish, do we have data about how much AEM instances this would serve?

File: lib/asset/createassetservletpart.js (hunk 2 above)
Comment (mfrisbey) [reply lib/asset/createassetservletpart.js:33]: @tmathern At least for VSCode intellisense, leaving the description blank on an overridden class method will inherit the parent method's description. This is why I've left the description missing.

File: lib/asset/transferpart.js (hunk 5 above)
Comment (mfrisbey) [reply lib/asset/transferpart.js:156]: It comes from the constants file, so AFAIK it will.

**Comment by mfrisbey** [comment]
> do we have data about how much AEM instances this would serve?

No concrete numbers. How would you hope to use that data?

From a technical perspective, here are some things we would gain with this change:

* `aem-upload` could be used to upload assets even to older AEM versions.
* As already mentioned, AEM Desktop could use a single dependency to upload files to AEM, instead of needing to maintain two implementations. The old implementation is less optimized, very complex, and based on the deprecated `request` module. I have a client project that will require updating this code so that it uses a more modern module like `fetch`, and this PR is exponentially simpler than trying to change the desktop app would be. 
* AFAIK, AAL is in a similar situation: it uses this library in some cases, but maintains its own, separate implementation for using the create asset servlet. This change would give us the option of simplifying that case, too.

With this change, I could deprecate a significant amount of code in AEM Desktop, improve its upload speeds against older AEM instances, and upgrade it to use newer, better supported technologies. Even without client numbers, the massive technical benefits of this feature alone would be more than worth it.

**Comment by mfrisbey** [comment]
Closing this PR for another with some improvements.

File: lib/asset/createassetservletpart.js (hunk 2 above)
Comment (tmathern) [reply lib/asset/createassetservletpart.js:33]: Not everyone uses VS Code, so I would suggest to still fill out the doc (at least add that it overrides the method from the parent).

**Comment by tmathern** [comment]
> > do we have data about how much AEM instances this would serve?
> 
> No concrete numbers. How would you hope to use that data?
> 
> From a technical perspective, here are some things we would gain with this change:
> 
> * `aem-upload` could be used to upload assets even to older AEM versions.
> * As already mentioned, AEM Desktop could use a single dependency to upload files to AEM, instead of needing to maintain two implementations. The old implementation is less optimized, very complex, and based on the deprecated `request` module. I have a client project that will require updating this code so that it uses a more modern module like `fetch`, and this PR is exponentially simpler than trying to change the desktop app would be.
> * AFAIK, AAL is in a similar situation: it uses this library in some cases, but maintains its own, separate implementation for using the create asset servlet. This change would give us the option of simplifying that case, too.
> 
> With this change, I could deprecate a significant amount of code in AEM Desktop, improve its upload speeds against older AEM instances, and upgrade it to use newer, better supported technologies. Even without client numbers, the massive technical benefits of this feature alone would be more than worth it.

It's to decide if it is worthwhile to do, maintain and refactor. If there still is a larger interest from that, it's worthwhile to invest time in refactoring/redesigning now. I understand the Desktop app's situation but we're introducing complexity here for systems we now consider legacy and try to move away from. This here leads to indirectly introducing it in 2 other products. So we need some special (engineering) care to make sure it's both usable and maintainable and that all products which would use the new version invest in doing the refactoring changes too.

**Pull request closed without merging**